presto-main/src/main/java/com/facebook/presto/execution/scheduler/LegacySqlQueryScheduler.java

```diff
@@ -459,9 +503,189 @@ else if (!result.getBlocked().isDone()) {
+        for (PlanOptimizer optimizer : runtimePlanOptimizers) {
+            newRoot = optimizer.optimize(newRoot, session, variableAllocator.getTypes(), variableAllocator, idAllocator, warningCollector);
+        }
+        // The partitioningScheme should stay the same
+        // even if the root's outputVariable layout is changed.
+        if (newRoot != fragment.getRoot()) {
+            oldToNewFragmentMapping.put(fragment, new PlanFragment(
+                    fragment.getId(),
+                    newRoot,
+                    fragment.getVariables(),
+                    fragment.getPartitioning(),
+                    scheduleOrder(newRoot),
+                    fragment.getPartitioningScheme(),
+                    fragment.getStageExecutionDescriptor(),
+                    fragment.isOutputTableWriterFragment(),
+                    fragment.getStatsAndCosts(),
+                    fragment.getJsonRepresentation()));
+        }
+    }
+
+    /**
+     * Utility function that rebuild a StreamingPlanSection, re-create stageExecutionAndScheduler for each of its stage, and finally update the stageExecutions map.
+     */
+    private void updateStageExecutions(StreamingPlanSection section, Map<PlanFragment, PlanFragment> oldToNewFragment)
+    {
+        StreamingPlanSection newSection = new StreamingPlanSection(rewriteStreamingSubPlan(section.getPlan(), oldToNewFragment), section.getChildren());
+        PlanFragment sectionRootFragment = newSection.getPlan().getFragment();
+        Optional<int[]> bucketToPartition;
+        OutputBuffers outputBuffers;
+        ExchangeLocationsConsumer locationsConsumer;
+        if (isRootFragment(sectionRootFragment)) {
+            bucketToPartition = Optional.of(new int[1]);
+            outputBuffers = createInitialEmptyOutputBuffers(sectionRootFragment.getPartitioningScheme().getPartitioning().getHandle())
+                    .withBuffer(new OutputBufferId(0), BROADCAST_PARTITION_ID)
+                    .withNoMoreBufferIds();
+            OutputBufferId rootBufferId = getOnlyElement(outputBuffers.getBuffers().keySet());
+            locationsConsumer = (fragmentId, tasks, noMoreExchangeLocations) ->
+                    updateQueryOutputLocations(queryStateMachine, rootBufferId, tasks, noMoreExchangeLocations);
+        }
+        else {
+            bucketToPartition = Optional.empty();
+            outputBuffers = createDiscardingOutputBuffers();
+            locationsConsumer = (fragmentId, tasks, noMoreExchangeLocations) -> {};
+        }
+        SectionExecution sectionExecution = sectionExecutionFactory.createSectionExecutions(
+                session,
+                newSection,
+                locationsConsumer,
+                bucketToPartition,
+                outputBuffers,
+                summarizeTaskInfo,
+                remoteTaskFactory,
+                splitSourceFactory,
+                0);
+        addStateChangeListeners(sectionExecution);
```

Comment:
do we need to remove the state change listeners for old sections?

Reply:
Yeah, I thought about this. When we created new stageExecutions and replace the old ones in `this.stageExecutions`, these old stageExecutions will have nothing referencing to them and will be garbage collected eventually without getting triggered. Therefore, I'm not too worried about explicitly removing listeners. Besides, stateMachine class also does not provide methods to explicitly remove listeners, so I think it is ok to trust on the GC to do its own job :)